#BN-27 feat: Recommended_Products

Illustre13 opened the pull request:
# Trello Board Link

- https://trello.com/c/vnDBEQU4

# PR Description
This PR will allow buyer to see recommended products
# Description of tasks that were expected to be completed.
The backend needs to provide recommendations to the frontend.
GIVEN a GET request for recommended products.

- [x] When the request is made by a buyer
- [x] Then a list of recommended products should be retrieved from the database
- [x] And sent to the frontend to be displayed

# Description of tasks that were expected to be completed and that are not yet completed
- [x] None
# How this may be tested?
On this task I have created dummy data for testing the algorithm that I come up with. To get started first,
1. Make sure you are having a local postgres database and server running on port 8080
2. Make sure you are having `ecommerce-app-cogito-db-dev` database or create a new one
3. Run `make migrate` to migrate all tables to your `ecommerce-app-cogito-app-dv` database
4. Run `make seed` to copy the dummy data to your tables in db
6. In Postman or Insomnia create request with `http://localhost:9090/products/recommended/:{buyer_id}` with a GET method.
7. With Swagger documentation, Go to `http://localhost:9090/api-docs/#/Products/get_products_recommended__id_` and pass in in the buyer id. You should first be logged in as the buyer in the system.
# Screenshots (If appropriate/applicable)
-[x] --
Please check this Checklist before you submit your PR:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My code generates no warnings
- [x] There are no vulnerabilities

Illustre13 committed:
feat(product endpoint): created crud endpoints for the product

Illustre13 committed:
chore(bug): fixed a bug

Illustre13 committed:
chore(merging): merged the develop branch to ft-recommended-products-27

Illustre13 committed:
chore(merging): merged the develop branch to ft-recommended-products-27

Illustre13 committed:
chore(merging): merged the develop branch to ft-recommended-products-27

Illustre13 committed:
feat(recommended products): working on recommended products

created a table productViews to record all the products the users have viewed

created productViews, user, and product dummy data for testing the recommended products endpoint

created a function to check all the viewed products by the user and return recommended products based on the user who have similar product references

EvelyneUmubyeyi commented on src/controllers/products/recommendedController.js:
```diff
@@ -0,0 +1,263 @@
+    fallbackLng: 'en',
+    preload: ['en', 'fr'],
+  });
+app.use(express.urlencoded({ extended: false }));
```
I think this block of code of initializing the app and translation configuration can stay in index.js. It doesn't have to be repeated here in the controller, you will still be able to send messages in French and English.

Illustre13 replied:
I have removed it from the controller.

EvelyneUmubyeyi commented on src/controllers/products/recommendedController.js:
```diff
@@ -0,0 +1,263 @@
+    const { authenticatedBuyer } = req;
+    if (!authenticatedBuyer) {
+      return res.status(403).json({
+        success: false,
```
I think it would be best to return response status code in the json instead of this success property

Illustre13 replied:
I think success is simply a parameter I included, it will not to cause any problem.

EvelyneUmubyeyi commented on src/controllers/products/recommendedController.js:
```diff
@@ -0,0 +1,263 @@
+    const products = await product.findAll();
+    res.status(200).json({
+      success: true,
+      message: req.t('Successfully retrieved all the products from the database'),
```
Looks like there are no translations for these messages in the locale folder

Illustre13 replied:
I have included translation on all my endpoints now!

EvelyneUmubyeyi commented on src/controllers/products/recommendedController.js:
```diff
@@ -0,0 +1,263 @@
+    });
+    res.status(200).json({
+      success: true,
+      message: `${req.t('Product Was Found with ID: ')} ${req.params.id} `,
```
the message returned only contains the id of the retrieved product

Illustre13 replied:
Have to check again, and the endpoint is working fine.

EvelyneUmubyeyi commented on src/controllers/products/recommendedController.js:
```diff
@@ -0,0 +1,263 @@
+  }
+}
+// function to get recommended product
+export const RecommendedProduct = async (req, res) => {
```
I have tested this function and it does well when a user has viewed products similar to other users as it goes ahead to recommend other products viewed by those users. However, when the user has not viewed any product similar to other users they don't get any recommendation. I would suggest that in that case, you recommend products in the same category as the product the user has already viewed. For instance, a user viewed a jacket but no other users viewed that jacket, so they could get recommendations of other products in the fashion category as it seems to be a category of interest.

Illustre13 replied:
Yes i'm still working to add other parameters while generating the recommended products. I will also include that.

Illustre13 committed:
fix(bug fix): worked on the review feedback from Evelyne

-added translation on my endpoint

Illustre13 committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-recommended-products-27

Illustre13 committed:
Merge branch 'develop' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-recommended-products-27

Illustre13 committed:
fix(bug fix): fixed bugs in my endpoint

added status code into the http response

Illustre13 committed:
test(recommended products): added test to the recommended products endpoint
- updated the recommended product algorithm

-added wishlist products into the recommended products

-added in the algorithm the products category of all products in the wishlist and productViews table

Illustre13 committed:
feat(recommended products): updated the recommended product algorithm

-added wishlist products into the recommended products

-added in the algorithm the products category of all products in the wishlist and productViews table

Illustre13 committed:
test(recommended products): added test to the recommended products endpoint

Illustre13 committed:
test(recommended products): added test to the recommended products endpoint

Illustre13 committed:
Merge branch 'ft-recommended-products-27' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-recommended-products-27

Illustre13 committed:
Merge branch 'ft-recommended-products-27' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-recommended-products-27

dushimeemma reviewed (commented):
Kindly swatch your commits, otherwise good job 🥇

Illustre13 committed:
Merge branch 'ft-recommended-products-27' of https://github.com/atlp-rwanda/ecommerce-app-cogito-bn into ft-recommended-products-27
